lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -195,3 +196,65 @@ SILInstruction *swift::findInitExistential(FullApplySite AI, SILValue Self,
   }
   return nullptr;
 }
+
+/// Derive a concrete type of self and conformance from the init_existential
+/// instruction.
+/// If successful, initializes a valid ConformanceAndConcreteType.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
+  // Try to find the init_existential, which could be used to
+  // determine a concrete type of the self.
+  // Returns: InitExistential, OpenedArchetype, OpenedArchetypeDef, isCopied.
+  InitExistential = findInitExistential(openedUse, OpenedArchetype,
+                                        OpenedArchetypeDef, isCopied);
+  if (!InitExistential)
+    return;
```

Comment:
@atrick @slavapestov Could you decouple the body from the constructor and invoke it separately via an API? "findInitExistential" is not the only way to determine concrete types. For example, I am using ProtocolConformance Analysis to determine concrete types. I am planning on using the same ConcreteExistentialInfo struct and reuse the createApplyWithConcreteType. Let me know if you are busy and it sounds good to you, I can try to decouple them.

Reply:
@rajbarik All suggestions are welcome. I'm not dogmatic about how ConcreteExistentialInfo is constructed or accessed. I like to start with minimal design and develop outward. If you have a rough prototype, and want quick feedback, just point me to a branch or gist.

First thing to consider: do all the members of ConcreteExistentialInfo make sense for all the SIL patterns you're handling. It's fine to have a few optional fields:
- ConcreteTypeDef is null to when ConcreteType is not an opened archetype
- You could make InitExistential null to indicate the existential wasn't found.

Anything much beyond that and it may be better to have a discrimiated union or a smaller more general Info composing a larger one.

At any rate, I think it's fine if you optionally pass in a PCA to the ConcreteExistentialInfo construction, or even define a second constructor. If that's awkward for some reason you could move to a free standing or static function to create the Info.

Reply:
@atrick Thanks. I am currently adding a new constructor for CEI with values computed from PCA and this constructor will take a subset of the values currently present in the CEI. We can later think about a more tighter integration. However, one problem I am running into is regarding the following code in method createApplyWithConcreteType:
```
if (CEI.isCopied && !canReplaceCopiedSelf(Apply, CEI.InitExistential, DA))
```
When we use PCA, there is no InitExistential and no isCopied. So, the question is, can this check be moved to isValid() function in CEI or prepended with a check on InitExistential not being null?  Do you have any recommendation?

Reply:
Please keep in mind that a check on InitExistential being not null, is minimal in terms of code changes :)

Reply:
The safe thing to do is to bail out under the `isCopied` case if we didn't find the init_existential.
`if (isCopied && (!InitExistential || !canReplaceCopiedSelf(Apply, InitExistential)`
 I'm sure we could do better but if you don't care about that case and never hit it then why deal with it?